### ISSUE TITLE
decorator for handlers

### ISSUE THREAD

**Issue by spaceone** [circuits/circuits#91]
We should enhance the possibility to write decorators for handler-functions, as `@handler` copies meta information from the function and does a arginspect if the first event is called 'event'.

We should provide a utility which developers can use so that they don't need to care about these internals when they want to write additional decorators.

Any further ideas on this? I have some ideas in mind which I will implement but I would like to heat what you think.

**Comment by prologic** [comment]
Hmm; I'm not quite sure what you mean? What ideas do you have in mind?

Do you mean a decorator to make writing custom decoraotrs that create event handlers a bit easier?

**Comment by spaceone** [comment]
The problem is that if you want to write a decorator which is applied after `@handler` one need to
* 1. copy the function metadata of the original function to the decorated function
* 2. copy the code which does the arginspect to know if one have to pass the 'event' to the next function.

**Comment by prologic** [comment]
Ahh I see; so what are you ideas then?

Turn ``@handler`` into a class that can be subclasses?

How else can decorators be extended in this way?

**Comment by prologic** [comment]
If I'm not mistaken I *think* you can do something like this:

```
def mydecorator(*args, **kwargS):
    @handler(...)
    def wrapper(f):
        return f(...)
    return wrapper
```

I think :)

**Comment by spaceone** [comment]
Ì think a class is not necessary?!
I would add a function `handler_decorator(func)` which does the necessary arginspect and use this instead of doing the arginspect in manager.py

**Comment by prologic** [comment]
e.g: [``circuits.web.controller.expose``](https://github.com/circuits/circuits/blob/master/circuits/web/controllers.py#L18)

**Comment by spaceone** [comment]
circuits.web.controller.expose is the best example for the redundant things i want to get rid of:
```
 if not getattr(f, "event", False):
    return f(self, *args, **kwargs)
else:
    return f(self, event, *args, **kwargs)
```

**Comment by prologic** [comment]
Yeah I agree; this is less to do with function parameter inspection and more to do with:

Do we need to send the ``event`` object?

I'm *not* sure how to solve this per se (yet)

**Comment by prologic** [comment]
Got any ideas around this? I haven't really looked much further into this (yet).
